src/ucp/proto/proto_am.inl

```diff
@@ -222,28 +222,66 @@ void ucp_dt_iov_copy_uct(ucp_context_h context, uct_iov_t *iov, size_t *iovcnt,
     state->offset += length_it;
 }
 
+static UCS_F_ALWAYS_INLINE
+ucs_status_t ucp_am_zcopy_common(ucp_request_t *req, const void *hdr,
+                                 size_t hdr_size, ucp_mem_desc_t *user_hdr_desc,
+                                 size_t user_hdr_size, uct_iov_t *iov,
+                                 size_t max_iov, size_t max_length,
+                                 uint8_t am_id, ucp_dt_state_t *state)
+{
+    ucp_ep_t *ep          = req->send.ep;
+    ucp_md_index_t md_idx = ucp_ep_md_index(ep, req->send.lane);
+    size_t iovcnt         = 0ul;
+    ucs_status_t status;
+
+    if (user_hdr_size != 0) {
+        ucs_assert(max_length > user_hdr_size);
+        ucs_assert(max_iov > 1);
+
+        iov[0].buffer = user_hdr_desc + 1;
+        iov[0].length = user_hdr_size;
+        iov[0].memh   = ucp_memh2uct(user_hdr_desc->memh, md_idx);
+        iov[0].stride = 0;
+        iov[0].count  = 1;
+
+        ucp_dt_iov_copy_uct(ep->worker->context, iov + 1, &iovcnt, max_iov - 1,
+                            state, req->send.buffer, req->send.datatype,
+                            max_length - user_hdr_size, md_idx, NULL);
+        ++iovcnt;
+    } else {
+        ucp_dt_iov_copy_uct(ep->worker->context, iov, &iovcnt, max_iov,
+                            state, req->send.buffer, req->send.datatype,
+                            max_length, md_idx, NULL);
```

Comment:
can we remove some code duplication by doing ++iov and --max_iov before calling ucp_dt_iov_copy_uct()?

Reply:
we can, but we will have to make a couple of unnecessary ops for empty header case. Imo, it looks ok - it just a single function invocation with different params.

Reply:
this function is inlined and if we invoke it 2 times it will be more code

Reply:
ok